.github/workflows/compiler_sanitizers.yml

```diff
@@ -76,54 +76,22 @@ jobs:
   clang_TSAN:
     # To enable this workflow on a fork, comment out:
     if: github.repository == 'numpy/numpy'
-    runs-on: macos-latest
+    runs-on: ubuntu-latest
+    container:
+        image: ghcr.io/nascheme/numpy-tsan:3.14t-dev
+        options: --shm-size=2g # increase memory for large matrix ops
+      
     steps:
-    - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
-      with:
-        submodules: recursive
-        fetch-tags: true
-        persist-credentials: false
-    - name: Set up pyenv
-      run: |
-        git clone https://github.com/pyenv/pyenv.git "$HOME/.pyenv"
-        PYENV_ROOT="$HOME/.pyenv"
-        PYENV_BIN="$PYENV_ROOT/bin"
-        PYENV_SHIMS="$PYENV_ROOT/shims"
-        echo "$PYENV_BIN" >> $GITHUB_PATH
-        echo "$PYENV_SHIMS" >> $GITHUB_PATH
-        echo "PYENV_ROOT=$PYENV_ROOT" >> $GITHUB_ENV
-    - name: Check pyenv is working
-      run:
-        pyenv --version
-    - name: Set up LLVM
-      run: |
-        brew install llvm@19
-        LLVM_PREFIX=$(brew --prefix llvm@19)
-        echo CC="$LLVM_PREFIX/bin/clang" >> $GITHUB_ENV
-        echo CXX="$LLVM_PREFIX/bin/clang++" >> $GITHUB_ENV
-        echo LDFLAGS="-L$LLVM_PREFIX/lib" >> $GITHUB_ENV
-        echo CPPFLAGS="-I$LLVM_PREFIX/include" >> $GITHUB_ENV
-    - name: Build Python with thread sanitizer support
-      run: |
-        # free-threaded Python is much more likely to trigger races
-        CONFIGURE_OPTS="--with-thread-sanitizer" pyenv install 3.13t
-        pyenv global 3.13t
-    - name: Install dependencies
-      run: |
-        # TODO: remove when a released cython supports free-threaded python
-        pip install -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple cython
-        pip install -r requirements/build_requirements.txt
-        pip install -r requirements/ci_requirements.txt
-        pip install -r requirements/test_requirements.txt
-        # xdist captures stdout/stderr, but we want the TSAN output
-        pip uninstall -y pytest-xdist
-    - name: Build
-      run:
-        python -m spin build -j2 -- -Db_sanitize=thread
```

Comment:
I'm pretty sure the failures you're seeing are caused by deleting this step. You need to explicitly compile NumPy with thread sanitizer.

Reply:
ok thanks! I'll make the adjustment